// @ui5/webcomponents-react-charts/dist/BulletChartPlaceholder.js
import { ThemingParameters } from '@ui5/webcomponents-react-base/dist/ThemingParameters';
import React from 'react';
import ContentLoader from 'react-content-loader';

var barContentVertical = /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("rect", {
  x: "20",
  y: "20",
  width: "85",
  height: "15"
}), /*#__PURE__*/React.createElement("rect", {
  x: "20",
  y: "40",
  width: "117",
  height: "15"
}), /*#__PURE__*/React.createElement("rect", {
  x: "20",
  y: "60",
  width: "67",
  height: "15"
}), /*#__PURE__*/React.createElement("rect", {
  x: "20",
  y: "80",
  width: "75",
  height: "15"
}), /*#__PURE__*/React.createElement("rect", {
  x: "20",
  y: "100",
  width: "125",
  height: "15"
}), /*#__PURE__*/React.createElement("rect", {
  x: "20",
  y: "120",
  width: "60",
  height: "15"
}), /*#__PURE__*/React.createElement("rect", {
  x: "115",
  y: "17.5",
  width: "2",
  height: "20"
}), /*#__PURE__*/React.createElement("rect", {
  x: "150",
  y: "37.5",
  width: "2",
  height: "20"
}), /*#__PURE__*/React.createElement("rect", {
  x: "90",
  y: "57.5",
  width: "2",
  height: "20"
}), /*#__PURE__*/React.createElement("rect", {
  x: "100",
  y: "77.5",
  width: "2",
  height: "20"
}), /*#__PURE__*/React.createElement("rect", {
  x: "120",
  y: "97.5",
  width: "2",
  height: "20"
}), /*#__PURE__*/React.createElement("rect", {
  x: "87",
  y: "117.5",
  width: "2",
  height: "20"
}));
var barContentHorizontal = /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("rect", {
  x: "20",
  y: "50",
  width: "15",
  height: "85"
}), /*#__PURE__*/React.createElement("rect", {
  x: "40",
  y: "70",
  width: "15",
  height: "65"
}), /*#__PURE__*/React.createElement("rect", {
  x: "60",
  y: "68",
  width: "15",
  height: "67"
}), /*#__PURE__*/React.createElement("rect", {
  x: "80",
  y: "60",
  width: "15",
  height: "75"
}), /*#__PURE__*/React.createElement("rect", {
  x: "100",
  y: "10",
  width: "15",
  height: "125"
}), /*#__PURE__*/React.createElement("rect", {
  x: "120",
  y: "75",
  width: "15",
  height: "60"
}), /*#__PURE__*/React.createElement("rect", {
  x: "17.5",
  y: "35",
  width: "20",
  height: "2"
}), /*#__PURE__*/React.createElement("rect", {
  x: "37.5",
  y: "62",
  width: "20",
  height: "2"
}), /*#__PURE__*/React.createElement("rect", {
  x: "57.5",
  y: "56",
  width: "20",
  height: "2"
}), /*#__PURE__*/React.createElement("rect", {
  x: "77.5",
  y: "37",
  width: "20",
  height: "2"
}), /*#__PURE__*/React.createElement("rect", {
  x: "97.5",
  y: "45",
  width: "20",
  height: "2"
}), /*#__PURE__*/React.createElement("rect", {
  x: "117.5",
  y: "70",
  width: "20",
  height: "2"
}));
var BulletChartPlaceholder = function BulletChartPlaceholder(_ref) {
  var layout = _ref.layout,
      measures = _ref.measures;
  var distinctElements = new Set(measures.map(function (item) {
    return item.type;
  }));
  return /*#__PURE__*/React.createElement(ContentLoader, {
    viewBox: "0 0 165 145",
    preserveAspectRatio: "xMidYMid meet",
    width: "100%",
    height: "100%",
    speed: 2,
    backgroundColor: ThemingParameters.sapContent_ImagePlaceholderBackground,
    foregroundColor: ThemingParameters.sapContent_ImagePlaceholderForegroundColor,
    backgroundOpacity: ThemingParameters.sapContent_DisabledOpacity
  }, layout === 'vertical' && /*#__PURE__*/React.createElement("rect", {
    x: "20",
    y: "10",
    width: "1",
    height: "135"
  }), layout === 'horizontal' && /*#__PURE__*/React.createElement("rect", {
    x: "10",
    y: "135",
    width: "135",
    height: "1"
  }), layout === 'vertical' && distinctElements.has('primary') && barContentVertical, layout === 'horizontal' && distinctElements.has('primary') && barContentHorizontal);
};

export { BulletChartPlaceholder };
