// @ui5/webcomponents-react-charts/dist/ScatterChartPlaceholder.js
import _extends from '@babel/runtime/helpers/extends';
import { ThemingParameters } from '@ui5/webcomponents-react-base/dist/ThemingParameters';
import React from 'react';
import ContentLoader from 'react-content-loader';

var ScatterChartPlaceholder = function ScatterChartPlaceholder(props) {
  return /*#__PURE__*/React.createElement(ContentLoader, _extends({
    viewBox: "0 0 165 145",
    preserveAspectRatio: "xMidYMid meet",
    width: "100%",
    height: "100%",
    speed: 2,
    backgroundColor: ThemingParameters.sapContent_ImagePlaceholderBackground,
    foregroundColor: ThemingParameters.sapContent_ImagePlaceholderForegroundColor,
    backgroundOpacity: ThemingParameters.sapContent_DisabledOpacity
  }, props), /*#__PURE__*/React.createElement("rect", {
    x: "10",
    y: "135",
    width: "135",
    height: "1"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "20",
    cy: "90",
    r: "2"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "40",
    cy: "58",
    r: "3"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "60",
    cy: "108",
    r: "14"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "85",
    cy: "87",
    r: "8"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "105",
    cy: "50",
    r: "6"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "125",
    cy: "115",
    r: "4"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "78",
    cy: "58",
    r: "10"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "28",
    cy: "115",
    r: "4"
  }));
};

export { ScatterChartPlaceholder };
