// @ui5/webcomponents-react-charts/dist/components/ChartContainer.js
import _extends from '@babel/runtime/helpers/extends';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _assertThisInitialized from '@babel/runtime/helpers/assertThisInitialized';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { ThemingParameters } from '@ui5/webcomponents-react-base/dist/ThemingParameters';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/dist/usePassThroughHtmlProps';
import { Label } from '@ui5/webcomponents-react/dist/Label';
import { Loader } from '@ui5/webcomponents-react/dist/Loader';
import React, { forwardRef, useMemo } from 'react';
import { createUseStyles } from 'react-jss';
import { ResponsiveContainer } from 'recharts';

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
// eslint-disable-next-line no-underscore-dangle
var __testingProps__ = {};

if (process.env.NODE_ENV === 'test') {
  __testingProps__.width = 400;
  __testingProps__.height = 400;
}

var loaderStyles = {
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0
};
var chartContainerStyles = {
  '@global': {
    '.has-click-handler': {
      '& .recharts-pie-sector, .recharts-bar-rectangles, .recharts-active-dot, .recharts-area-dot': {
        cursor: 'pointer'
      }
    }
  }
};
var useStyles = createUseStyles(chartContainerStyles, {
  name: 'ChartContainer'
});

var ErrorBoundary = /*#__PURE__*/function (_React$Component) {
  _inherits(ErrorBoundary, _React$Component);

  var _super = _createSuper(ErrorBoundary);

  function ErrorBoundary() {
    var _this;

    _classCallCheck(this, ErrorBoundary);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      errorCount: 0
    });

    return _this;
  }

  _createClass(ErrorBoundary, [{
    key: "componentDidCatch",
    value: function componentDidCatch() {
      if (this.state.errorCount < 3) {
        this.setState(function (old) {
          return _objectSpread(_objectSpread({}, old), {}, {
            errorCount: old.errorCount + 1
          });
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      if (this.state.errorCount >= 3) {
        return /*#__PURE__*/React.createElement(Label, null, "Sorry, something went wrong while rendering this chart!");
      }

      return this.props.children;
    }
  }]);

  return ErrorBoundary;
}(React.Component);

var ChartContainer = /*#__PURE__*/forwardRef(function (props, ref) {
  var Placeholder = props.Placeholder,
      _props$loading = props.loading,
      loading = _props$loading === void 0 ? false : _props$loading,
      dataset = props.dataset,
      style = props.style,
      className = props.className,
      tooltip = props.tooltip,
      slot = props.slot,
      children = props.children,
      resizeDebounce = props.resizeDebounce;
  useStyles();
  var internalStyles = useMemo(function () {
    return _objectSpread({
      fontSize: ThemingParameters.sapFontSmallSize,
      color: ThemingParameters.sapTextColor,
      fontFamily: ThemingParameters.sapFontFamily,
      width: '100%',
      height: '400px',
      position: 'relative'
    }, style !== null && style !== void 0 ? style : {});
  }, [style]);
  var passThroughProps = usePassThroughHtmlProps(props);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    style: internalStyles,
    className: className,
    title: tooltip,
    slot: slot
  }, passThroughProps), (dataset === null || dataset === void 0 ? void 0 : dataset.length) > 0 ? /*#__PURE__*/React.createElement(React.Fragment, null, loading && /*#__PURE__*/React.createElement(Loader, {
    style: loaderStyles
  }), /*#__PURE__*/React.createElement(ErrorBoundary, null, /*#__PURE__*/React.createElement(ResponsiveContainer, _extends({
    debounce: resizeDebounce
  }, __testingProps__), children))) : /*#__PURE__*/React.createElement(Placeholder, null));
});
ChartContainer.displayName = 'ChartContainer';

export { ChartContainer };
